File: dp/longest_palindrome_string.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
 public:
  string longestPalin(string S) {
    // code here
    string ans;
    int len = INT_MIN;
    vector<vector<int>> dp(S.size(), vector<int>(S.size()));
    for (int i = 0; i < S.size(); ++i) dp[i][i] = 1;
    for (int j = 1; j < S.size(); ++j) {
      if (S[j] == S[j - 1]) {
        dp[j - 1][j] = 1;
      }
    }
    for (int j = 2; j < S.size(); ++j) {
      for (int i = 0; i + j < S.size(); ++i) {
        if (dp[i + 1][i + j - 1] == 1 && S[i] == S[i + j]) {
          dp[i][i + j] = 1;
        }
      }
    }
    for (int i = 0; i < S.size(); ++i) {
      for (int j = i; j < S.size(); ++j) {
        if (dp[i][j] == 1) {
          // cout << i << ' ' << j << endl;
          if (j - i + 1 > len) {
            len = j - i + 1;
            ans = S.substr(i, j - i + 1);
          }
        }
      }
    }
    return ans;
  }
};

//{ Driver Code Starts.

int main() {
  int t;
  cin >> t;
  while (t--) {
    string S;
    cin >> S;

    Solution ob;
    cout << ob.longestPalin(S) << endl;
  }
}
// Contributed By: Pranay Bansal

// } Driver Code Ends

File: stl/longest_palindrome_string.py
# User function Template for python3


class Solution:
    def manacher_odd(self, s):
        n = len(s)
        s = "$" + s + "^"
        p = [0] * (n + 2)
        l, r = 1, 1
        for i in range(1, n + 1):
            p[i] = max(0, min(r - i, p[l + (r - i)]))
            while s[i - p[i]] == s[i + p[i]]:
                p[i] += 1
            if i + p[i] > r:
                l, r = i - p[i], i + p[i]
        return p[1 : n + 1]

    def longestPalin(self, S):
        # code here
        s = "#" + "#".join(c for c in S) + "#"
        p = self.manacher_odd(s)
        ma = -1
        idx = -1
        for i in range(len(p)):
            if p[i] > ma:
                ma = p[i]
                idx = i

        ans = s[idx - ma + 1 : idx + ma - 1]
        return "".join(ans.split("#"))


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        S = input()

        ob = Solution()

        ans = ob.longestPalin(S)

        print(ans)
# } Driver Code Ends
